### PR TITLE
feat(mypy): Hint for files existing outside project directory

### PR DESCRIPTION
**Pull request by captain-kark**
I am working in a project where all tests are placed in a sibling directory
next to the main project's source root.

```
cd ./some_project
tree
├── Makefile
├── README.md
├── some_project
│   ├── __init__.py
│   └── __main__.py
├── mypy.ini
├── pytest.ini
├── requirements-frozen.txt
├── requirements.txt
└── tests
    ├── conftest.py
    └── unit_tests
        ├── some_feature
        │   ├── test_foo.py
        │   └── test_bar.py
        └── test_baz.py
```

If you include a `mypy_path` in `mypy.ini` pointing to your current project,
when combined with `find_project_root`, a check can be made to use your
project's existing mypy type definitions while editing code in the `tests`
directory.

When editing code that matches a directory that shares its base with
a `mypy_path`, nothing is done since it is already in mypy's context.

### CODE REVIEW

File: bin/pycheckers.py [hunk 1]
```diff
@@ -766,6 +766,15 @@ def get_run_flags(self, filepath):
         config_file = self.find_config_file('mypy_config_file', ['mypy.ini'])
         if config_file:
             flags += ['--config-file', config_file]
+            mypy_config = ConfigParser()
+            mypy_config.read(config_file)
+            if mypy_config.has_option('mypy', 'mypy_path'):
+                # contextual source for mypy, useful when tests are outside module definition
+                for mypy_path in mypy_config.get('mypy', 'mypy_path').replace(':', ',').split(','):
```
Comment (msherry) [inline bin/pycheckers.py:773]: Why `.replace(':', ',').split(',')` instead of `.split(':')`?

File: bin/pycheckers.py [hunk 2]
```diff
@@ -766,6 +766,15 @@ def get_run_flags(self, filepath):
...
+                    try:
+                        os.path.dirname(original_filepath).index(os.path.join(os.path.dirname(config_file), mypy_path))
```
Comment (msherry) [inline bin/pycheckers.py:775]: Can you add a comment describing what the intent of this line is?

File: bin/pycheckers.py [hunk 3]
```diff
@@ -800,6 +806,29 @@ def get_filepath(self, filepath):
...
+            for mypy_path in mypy_config.get('mypy', 'mypy_path').replace(':', ',').split(','):
+                mypy_paths.append(os.path.join(config_dir, mypy_path))
+        return mypy_paths
+
+    def is_directory_outside_mypy_path(self, original_filedir, mypy_path):
+        # type: (str, str) -> bool
+        """Determine if `original_filedir` exists outside of `mypy_path`'s directory.
+        """
+        try:
+            original_filedir.index(mypy_path)
```
Comment (msherry) [inline bin/pycheckers.py:828]: Can we fix this up to use `os.path.commonprefix` or similar, to avoid issues on weirdly-nested paths? It's a small unlikely edge case, but I'd rather keep it clean from the start, if we can.

File: bin/pycheckers.py [hunk 4]
```diff
@@ -766,6 +766,12 @@ def get_run_flags(self, filepath):
         config_file = self.find_config_file('mypy_config_file', ['mypy.ini'])
         if config_file:
             flags += ['--config-file', config_file]
+            original_filedir = os.path.dirname(original_filepath)
+            for mypy_path in self.get_mypy_paths(config_file):
+                if self.is_directory_outside_mypy_path(original_filedir, mypy_path):
+                    # contextual source for a mypy module
+                    # i.e., when a module's tests are outside the module's definitions
+                    flags.append(mypy_path)
```
Comment (msherry) [inline bin/pycheckers.py:774]: Oof, `flags` is increasingly misnamed. Not your fault, just something I'm seeing =/

**Comment by captain-kark** [comment]
As it stands, right now this pull request will only suppress errors where mypy can't find a module definition for your current project's module, but type checking isn't working. I'm investigating a couple of issues with my setup/project and determining which of the following could be causing issues with the behavior I'm seeing versus what I would expect:

  - my `some_project` module only contains an init/main python file and isn't type checking anything at all unless I start passing arguments into the top-level function I am invoking immediately after checking `if __name__ == '__main__'`.

  - my tests directory is able to import a valid `some_project` method member and detect issues _from pylint_, but mypy still isn't complaining when I, for instance, pass a single int to a function signature requiring a str.

  - moving functions into a `lib.py` outside of `__init__.py` isn't helping

I'll be making a `some_project/lib/__init__.py` module and repeat all of these steps, and next. If that doesn't work, I'll drop the `__main__.py` entirely and move everything out of `__init__.py` so that it only imports from sibling module definitions. If there still isn't any progress on getting type checking working in my tests, I'll look into either moving my tests into the module source directly, and later, seeing if there's any special edge cases needed when using mypy in a directory that's expected to host pytest runners.

Any of this ringing a bell for you? Any ideas what I might be doing wrong, or why I might be having issues getting mypy to return false positives like this?
